src/mame/drivers/mc10.cpp

```diff
@@ -26,123 +24,102 @@
 ***************************************************************************/
 
-class mc10_state : public driver_device
+class mc10_state : public driver_device, public device_mc10cart_host_interface
 {
 public:
 	mc10_state(const machine_config &mconfig, device_type type, const char *tag)
 		: driver_device(mconfig, type, tag)
 		, m_maincpu(*this, "maincpu")
 		, m_ram(*this, RAM_TAG)
-		, m_bank1(*this, "bank1")
-		, m_bank2(*this, "bank2")
+		, m_mc10cart(*this, "ext")
 		, m_mc6847(*this, "mc6847")
-		, m_ef9345(*this, "ef9345")
 		, m_dac(*this, "dac")
 		, m_cassette(*this, "cassette")
 		, m_rs232(*this, "rs232")
 		, m_pb(*this, "pb%u", 0U)
-		{ }
+		{}
 
-	void alice90(machine_config &config);
-	void alice32(machine_config &config);
+	void mc10_base(machine_config &config);
+	void mc10_video(machine_config &config);
 	void mc10(machine_config &config);
+	void alice(machine_config &config);
+	void mc10_bfff_w_dac(uint8_t data);
+	uint8_t mc10_bfff_r();
 
 protected:
-	uint8_t mc10_bfff_r();
 	void mc10_bfff_w(uint8_t data);
 
 	uint8_t mc10_port1_r();
 	void mc10_port1_w(uint8_t data);
 	uint8_t mc10_port2_r();
 	void mc10_port2_w(uint8_t data);
-	uint8_t alice90_bfff_r();
-	void alice32_bfff_w(uint8_t data);
 
 	uint8_t mc6847_videoram_r(offs_t offset);
-	TIMER_DEVICE_CALLBACK_MEMBER(alice32_scanline);
 
 	// device-level overrides
 	virtual void driver_start() override;
 	virtual void driver_reset() override;
 
 	required_device<m6803_cpu_device> m_maincpu;
 	required_device<ram_device> m_ram;
-	required_memory_bank m_bank1;
-	optional_memory_bank m_bank2;
+	optional_device<mc10cart_slot_device> m_mc10cart;
 
-	uint8_t *m_ram_base;
-	uint32_t m_ram_size;
 	uint8_t m_keyboard_strobe;
 	uint8_t m_port2;
 
 	uint8_t read_keyboard_strobe(bool single_line);
 
+	mc10cart_slot_device &mc10cart() { return *m_mc10cart; }
+
 private:
-	void alice32_mem(address_map &map);
-	void alice90_mem(address_map &map);
 	void mc10_mem(address_map &map);
 
 	optional_device<mc6847_base_device> m_mc6847;
-	optional_device<ef9345_device> m_ef9345;
 	required_device<dac_bit_interface> m_dac;
 	required_device<cassette_image_device> m_cassette;
 	required_device<rs232_port_device> m_rs232;
 	required_ioport_array<8> m_pb;
 };
 
-class mcx128_state : public mc10_state
+class alice32_state : public mc10_state
 {
 public:
-	mcx128_state(const machine_config &mconfig, device_type type, const char *tag)
-		: mc10_state(mconfig, type, tag)
-		, m_mcx_ram(*this, "mcx_ram")
-		, m_mcx_cart_rom_base(*this, "cart")
-		, m_mcx_int_rom_base(*this, "maincpu")
-		, m_bank3(*this, "bank3")
-		, m_bank4r(*this, "bank4r")
-		, m_bank4w(*this, "bank4w")
-		, m_bank5r(*this, "bank5r")
-		, m_bank5w(*this, "bank5w")
-		, m_bank6r(*this, "bank6r")
-		, m_bank6w(*this, "bank6w")
-		{ }
-
-	void mcx128(machine_config &config);
+	alice32_state(const machine_config &mconfig, device_type type, const char *tag)
+	: mc10_state(mconfig, type, tag)
+	, m_ef9345(*this, "ef9345")
+	{}
```

Comment:
Indentation’s off on this, and a number of other initialiser lists – it should be like
```
    class_name(params...)
        : base_ame(args...)
        , member(args....)
        , ...
    { }
```